design/src/rfcs/rfc0037_retry_classifier_customization.md

```diff
@@ -0,0 +1,240 @@
+	    // preceding classifier and forward it on.
+		if let Some(result) = result_of_preceding_classifier {
+			return result;
+		}
+
+		todo!("inspect the interceptor context to determine if a retry attempt should be made.")
+	}
+
+	fn name(&self) -> &'static str { "my custom retry classifier" }
+}
+```
+### Customizing retry classification for a service
+
+```rust
+#[tokio::main]
+async fn main() -> Result<(), aws_sdk_s3::Error> {
+    let sdk_config = aws_config::load_from_env().await;
+    let service_config = aws_sdk_s3::Config::from(&sdk_config)
+	    .to_builder()
+	    .retry_classifier(CustomRetryClassifier)
+	    .build()
+    let client = aws_sdk_s3::Client::from_conf(&service_config);
+
+    let res = client
+	    .list_buckets()
+	    .send()
+	    .await?;
+
+	println!("your buckets: {res:?}");
+
+    Ok(())
+}
+```
+
+### Customizing retry classification for an operation
+
+```rust
+#[tokio::main]
+async fn main() -> Result<(), aws_sdk_s3::Error> {
+    let sdk_config = aws_config::load_from_env().await;
+    let client = aws_sdk_s3::Client::new(&sdk_config);
+
+    let res = client
+	    .list_buckets()
+        .customize()
+        .await
+        .unwrap()
+        .config_override(   
+            aws_sdk_s3::Config::builder()
+	            .retry_classifier(CustomRetryClassifier)
+        )
+        .send()
+        .await?;
+
+	println!("your buckets: {res:?}");
+
+    Ok(())
+}
+```
+
+## How to actually implement this RFC
+
+In order to implement this feature, we must:
+- Update the current retry classification system so that individual classifiers as well as collections of classifiers can be easily composed together.
+- Create two new configuration mechanisms for users that allow them to customize retry classification at the service level and at the operation level.
+- Update retry classifiers so that they may 'short-circuit' the chain, ending retry classification immediately.
+
+### The `RetryClassifier` trait
+
+```rust
+/// The result of running a [`ClassifyRetry`] on a [`InterceptorContext`].
+#[non_exhaustive]
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum RetryClassifierResult {
+    /// "A retryable error was received. This is what kind of error it was,
+    /// in case that's important."
+    Error(ErrorKind),
+    /// "The server told us to wait this long before retrying the request."
+    Explicit(Duration),
+    /// "This response should not be retried."
+    DontRetry,
+}
+
+/// Classifies what kind of retry is needed for a given [`InterceptorContext`].  
+pub trait ClassifyRetry: Send + Sync + fmt::Debug {
+    /// Run this classifier on the [`InterceptorContext`] to determine if the previous request
+    /// should be retried. Returns a [`RetryClassifierResult`].  
+    fn classify_retry(
+        &self,
+        ctx: &InterceptorContext,
+        result_of_preceding_classifier: Option<RetryClassifierResult>,
+    ) -> Option<RetryClassifierResult>;
```

Comment:
What does returning `None` mean? What happens if all the classifiers return `None`?

Should retry classifiers by fallible?

Reply:
I'll clarify.

Reply:
`None` means that a classifier had nothing to say. For example, this would happen when a classifier that relies on successful response deser encounters a deser failure. If all return `None`, then no retry is necessary